cc32xx watchdog support

locomuco opened the pull request:
this PR adds basic watchdog support for the cc32xx platforms to zephyr

locomuco committed:
cc32xx: add watchdog support for cc32xx devices

Signed-off-by: Matthias Boesl <matthias.boesl@gmail.com>

tbursztyka commented on drivers/watchdog/wdt_cc32xx.c:
```diff
@@ -0,0 +1,116 @@
+	if (user_cb != NULL) {
+		user_cb(wdog_r);
+	}
+	while(1);
```
add an empty line after } unless it's another } etc..

locomuco replied:
in general, is there a clang-format file available? where can I find style rules?

tbursztyka commented on drivers/watchdog/wdt_cc32xx.c:
```diff
@@ -0,0 +1,116 @@
+static void wdt_cc32xx_reload(struct device *dev)
+{
+	ARG_UNUSED(dev);
+	MAP_WatchdogReloadSet(WDT_BASE, CONFIG_WDT_CC32XX_RELOAD_TIME * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
```
We enforce the rule of 80chars limit so apply this rule everywhere.

tbursztyka commented on drivers/watchdog/wdt_cc32xx.c:
```diff
@@ -0,0 +1,116 @@
+	/* store ref for callback */
+	wdog_r = dev;
+	
+    MAP_PRCMPeripheralClkEnable(PRCM_WDT, PRCM_RUN_MODE_CLK);
```
tab for indentation

tbursztyka commented on drivers/watchdog/wdt_cc32xx.c:
```diff
@@ -0,0 +1,116 @@
+	MAP_WatchdogIntClear(WDT_BASE);
+
+	IRQ_CONNECT(EXCEPTION_WDT0, CONFIG_WDT_CC32XX_IRQ_PRI,
+		    	wdt_cc32xx_isr, DEVICE_GET(wdt_cc32xx), 0);
```
parameter alignement: it should be below the last '(' + 1 space. The indentation is thus made of tab(s) and extra spaces if necessary.

tbursztyka reviewed (changes requested):
small style issues.

tbursztyka commented on drivers/watchdog/wdt_cc32xx.c:
```diff
@@ -0,0 +1,116 @@
+
+static void wdt_cc32xx_disable(struct device *dev)
+{
+
```
you could add ARG_UNUSED(DEV) here.

tbursztyka commented on drivers/watchdog/Kconfig.cc32xx:
```diff
@@ -0,0 +1,38 @@
+
+config WDT_CC32XX_RELOAD_TIME
+	int "Watchdog timer reload in sec"
+	depends on WDT_CC32XX
```
instead of using depends on WDT_CC32XX, in the file dedicated for WDT_CC32XX, just enclose the relevant config options by 
if WDT_CC32XX

config ...

endif # WDT_CC32XX

(do not forget to break a line after that last endif, Kconfig parser is a bit dumb on this)

After the menuconfig of course. 

It keeps depends on for external dependencies, as well as reducing the file by some lines as well.

locomuco commented:
@tbursztyka all changes applied

GAnthony commented:
Thanks for adding this support to cc32xx!
A few comments:
* Tried to build samples/driver/watchdog for cc3220sf_launchxl, but got some errors.
   - Note: cc3200 is being deprecated, in favour of cc3220sf.  Could we make it build for the cc3220sf?
   - Need to define WDT_0_NAME (in drivers/watchdog/Kconfig.cc32xx or better yet in a dts watchdog entry)

locomuco commented:
@GAnthony what's the reason for deprecating the cc3200?

Sure I can take care about the other 2 changes..

GAnthony commented:
Well, TI prefers users move to the newer cc3220sf, given it has more flash memory and updated WiFi firmware (plus more security).

locomuco commented:
got it, but what's the downside of leaving it inside zephyr?

GAnthony commented:
I guess just the extra effort to support/test two essentially similar boards, but one is really meant to be a replacement of the other.

locomuco commented:
@GAnthony changes applied, works for both boards now

galak commented:
@GAnthony look good?

GAnthony commented:
I fetched the PR:
% git fetch upstream pull/559/head:cc32xx-wdt
% git checkout cc32xx-wdt
 and built samples/drivers/watchdog for the cc3220sf_launchxl.

Running via gdb, I get:
```
Start watchdog test
***** MPU FAULT *****
  Executing thread ID (thread): 0x20002a54
  Faulting instruction address:  0xc400c8f2
  Instruction Access Violation
Fatal fault in essential thread! Spinning...
```
This was because:
`
wdt_dev = device_get_binding("WATCHDOG_0");`

is returning NULL, so the sample crashes on dereference.
I think the CONFIG_WDT_0_NAME needs to be passed to DEVICE_AND_API_INIT instead of CONFIG_WDT_CC32XX_DEVICE_NAME for the sample to work.

Also, noted that WDT_CC32XX is on by default.  I'm wondering if it should be off by default, as not all applications may need it?

(Am I pulling the wrong code?)

locomuco commented:
@GAnthony changes applied

GAnthony commented:
I tried again the samples/drivers/watchdog sample (after enabling CONFIG_WDT_CC32XX) and though I did get a WDT interrupt once; thereafter, even after powering down and resetting the board, I get crashes.

Recommend looking at the SimpleLink SDK's watchdog timer driver for clues.   
One thing I see different in SimpleLink SDK, is the enabling of the WDT clock in Sleep mode also, like this:

            MAP_PRCMPeripheralClkEnable(id,
                PRCM_RUN_MODE_CLK | PRCM_SLP_MODE_CLK);

Maybe that would prevent the crashes?

Can you please test the patch on a CC3220SF launchpad before resubmitting?
Thnx!

locomuco commented:
@GAnthony can I ask you for support on this, I don't have a TI lauchpad for the CC3200sf on hand?

GAnthony commented:
Sure, but did this work on the CC3200 launchpad?  What is the need for having this driver?

Commit:
cc32xx: watchdog fixes and added debugger stall option

This patch fixes basic watchdog timer support.

True watchdog reset support is requires processing the
WDT_MODE configuration argument to wdt_set_config().

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

GAnthony commented:
@locomuco , managed to get watchdog working as a timer interrupt without reset:
See:
https://github.com/GAnthony/zephyr/commit/21893ed299a1ce0c5875860be5356cb8f585449f

I notice the Zephyr watchdog sample and driver test is based on Intel Quark, and the WDT APIs are not documented.  But it seems to get  reset functionality one needs to process the WDT_MODE paramater, and then not clear the WDT interrupt during the ISR.

I added a debugger stall option also, which cc32xx supports.

So, this is basically functional, but needs the work I mentioned (process WDT_MODE, avoid clearing interrupt in ISR), to get reset working.

Feel free to take the changes into your PR.

galak commented:
Any updates on this?

locomuco commented:
@galak  added requested changes from @GAnthony

GAnthony commented:
Did a forced push of the commit to do more cleanup of this driver, here
https://github.com/GAnthony/zephyr/commit/b1ac1ed46f02c03c20bbcecab521211784f704ad

However, after many attempts, still could not get it working on the CC3220SF.   After enabling the WDT, just accessing the WDT registers seems to disable the XDS110 debugger intermittently.

@locomuco, since there is no need for this driver, and it's not a priority, unless you actually intend to validate this driver on actual hardware, I recommend closing this PR for now.

galak commented:
@GAnthony should we just close this out?

GAnthony commented:
Yes, best to close it.
